src/compiler/transformers/es2015.ts

```diff
@@ -1344,6 +1344,38 @@ namespace ts {
             return node && node.dotDotDotToken && node.name.kind === SyntaxKind.Identifier && !inConstructorWithSynthesizedSuper;
         }
 
+        /**
+         * Gets the index of the first statement that contains a reference to a given parameter
+         *
+         * @param restParameter The ParameterDeclaration to which we are finding references.
+         * @param statements An array of Statement nodes to search
+         */
+        function findRestParameterReferenceIndex(restParameter: ParameterDeclaration, statements: Statement[]) {
+            const originalRestParameter = getOriginalNode(restParameter);
+
+            return findIndex(statements, containsReferenceToParameter);
+
+            function containsReferenceToParameter(statement: Statement): boolean {
+                return (isIdentifier(statement) && originalRestParameter === resolver.getReferencedValueDeclaration(statement)) ||
+                    forEachChild(statement, containsReferenceToParameter);
+            }
+        }
+
+        /**
+         * Gets the index of the first statement that contains a re-definition of "arguments"
+         *
+         * @param statements An array of Statement nodes to search
+         */
+        function findArgumentsRedefinitionIndex(statements: Statement[]) {
+            return findIndex(statements, containsArgumentsRedeclaration);
+
+            function containsArgumentsRedeclaration(statement: Statement): boolean {

```

Comment:
As above, this is yet another expensive recursive walk of the entire subtree and is better handled in the checker.